src/app/utils/pickupTimeEstimator.js

```diff
@@ -3,178 +3,445 @@ import nyplCoreObjects from '@nypl/nypl-core-objects'
-	// TO DO: what happens if the library is closed for a week?
-}
-
-export const _operatingHours = async (deliveryLocation) => {
-	const client = await nyplApiClient()
-	let hours
-	// if cache is less than one hour old, it is still valid
-	if (cache.hours && Date.now() - cache.updatedAt < 60 * 60 * 1000) {
-		hours = await cache.hours
-	}
-	else {
-		cache.hours = client.get(`/locations?location_codes=${deliveryLocation}&fields=hours`)
-		cache.updatedAt = Date.now()
-		hours = await cache.hours
-	}
-	if (!hours || !hours[deliveryLocation] || !hours[deliveryLocation][0] || !hours[deliveryLocation][0].hours) {
-		return []
-	}
-	return hours[deliveryLocation][0].hours
-}
-
-export const _resetCacheForTesting = (time = null) => {
-	cache.updatedAt = time
-	cache.hours = undefined
+/**
+ *  Given a DiscoverApi item, an optional delivery location id, and a optional request timestamp, returns an object defining:
+ *   - time: An ISO8601 formatted timestamp represeting the estimated time of arrival at the holdshelf
+ *   - estimate: A "friendly" statement built from the estimated time
+ *
+ *  The process for estimating arrival-at-holdshelf is:
+ *   - Compute arrivalAtDestination as:
+ *     - If onsite, it's now
+ *     - If offsite:
+ *       - Determine earliest processing time at origin (now if currently open; otherwise start of next operating hours)
+ *       - Add depository specific travel time
+ *   - Compute destinationServiceTime as:
+ *     - Determine earliest processing time at destination starting at arrivalAtDestination
+ *   - Compute arrivalAtHoldshelf as:
+ *     - Add location specific onsite travel time to destinationServiceTime
+ *     - Bump to next delivery time for rooms with special delivery schoedules
+ */
+estimator.getPickupTimeEstimate = async (item, deliveryLocationId, fromTimestamp = estimator._now()) => {
+  const fulfillment = fulfillments[item.physFulfillment]
+
+  // If no deliveryLocation specified, fall back on the one associated with the
+  // fulfillment or just 'ma' (which is a fine default for offsite requests)
+  deliveryLocationId = deliveryLocationId || fulfillment.location || 'ma'
+
+  const rationale = [{ time: fromTimestamp, activity: 'request time' }]
+
+  const originLocationId = estimator._locationId(item)
+
+  // Assume onsite request (i.e. already arrived at "destination" building)
+  let arrivalAtDestination = fromTimestamp
+
+  // If offsite:
+  if (/^rc/.test(originLocationId)) {
+    const originServiceTime = await estimator._getServiceTime(originLocationId, fromTimestamp)
+
+    rationale.push({ time: originServiceTime, activity: 'origin service time' })
+
+    // Calculate travel time as the fulfillment duration as the stated duration
+    // (e.g. PT2D, PT1D) minus 12H (artificially reduced to translate the colloquial
+    // "T2D" into the more accurate "T1D12H")
+    const offsiteTravelDuration = toSeconds(parseDuration(fulfillment.estimatedTime)) / 60 - 12 * 60
```

Comment:
can you add parens or line breaks so the order of operations is a little clearer?

Reply:
I put this admittedly confusing logic into its own function with improved comments.